src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,120 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double GROW_INDEX = 1.5;
+    private static final int START_INDEX_VALUE = 0;
+    private T[] dataArray;
+    private int size;
+
+    public ArrayList() {
+        dataArray = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
+    public ArrayList(int capacity) {
+        dataArray = (T[]) new Object[capacity];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == dataArray.length) {
+            ensureCapacity(size);
+        }
+        dataArray[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("index is invalid");
+        }
+        if (size == dataArray.length) {
+            ensureCapacity(size);
+        }
+        if (index == size) {
+            dataArray[size++] = value;
+        } else {
+            System.arraycopy(dataArray, index, dataArray, index + 1, size - index);
+            dataArray[index] = value;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        int neededSize = size + list.size();
+        int indexOfValueFromList = START_INDEX_VALUE;
+        if (dataArray.length < neededSize) {
+            ensureCapacity(neededSize);
+        }
+        for (int i = size; i < neededSize; i++) {
+            dataArray[i] = list.get(indexOfValueFromList++);
+            size++;
+        }
```

Comment:
just reuse method add

Reply:
In 'addAll' method I provided ensuring of capacity to be performed only once via checking:
```
if (dataArray.length < neededSize) {
            ensureCapacity(neededSize);
        }
```
This avoids repeated checks and expansions of the array.

If I reuse the add method, the program will redundantly check whether 'enshureCapacity' is needed every time the add method is executed:
```
public void add(T value) {
        if (size == dataArray.length) {
            ensureCapacity(size);
        }
        dataArray[size++] = value;
    }
```
This is why I add elements directly to the dataArray instead of reusing method add.

Or am I wrong and I still need to reuse the add method.

Reply:
You are right, but the performance impact of `if` is totally negligible, i'd argue having a clear and concise code is more important in most cases